### PR TITLE
Remove test-generator dependency 

### PR DESCRIPTION
**Pull request by PRATHAM-SPS**
### Details:
 - *Replaced the generator's functionality with unittest's subTest*

### Tickets:
 - *110705*

### CODE REVIEW

**Review by p-wysocki — COMMENTED** [review]
Thank you for your contribution!

Could you also remove `test-generator` from `openvino/tools/constraints.txt` and `openvino/tools/mo/requirements_dev.txt`?

**Review by jiwaszki — APPROVED** [review]
Removal LGTM, only requirements files need adjustments.

**Comment by PRATHAM-SPS** [comment]
Hii @p-wysocki  actually there are more 62 files are remained to modify. Once all files are changed I will remove it from openvino/tools/constraints.txt and openvino/tools/mo/requirements_dev.txt too.

**Comment by p-wysocki** [comment]
> Or should I remove it now only?

I didn't realize that there were more files - in that case you can do it whenever you're fine with it, there's no rush with its removal. :)

**Comment by PRATHAM-SPS** [comment]
Hii @p-wysocki @jiwaszki  @rkazants  where should I commit my new changes? In this branch olny or should I create New one?

**Comment by p-wysocki** [comment]
Please commit to this branch, everything will be in one place. :)

**Comment by rkazants** [comment]
@p-wysocki, please do not merge. Mo team needs to review it first to make sure no test coverage is lost

**Comment by p-wysocki** [comment]
@PRATHAM-SPS let's worry about it once everything is done, we currently have some CI issues so let's skip that topic for now. :)

**Comment by PRATHAM-SPS** [comment]
Hii @p-wysocki @jiwaszki @rkazants Task is completed PR is ready to merge.

**Comment by p-wysocki** [comment]
Looks great! There are a few failing MO tests in `openvino-lin` check in the precommit CI - could you please check if you have access to the logs?

**Comment by p-wysocki** [comment]
Great!

> @p-wysocki, please do not merge. MO team needs to review it first to make sure no test coverage is lost
@rkazants could you please take a look

**Comment by PRATHAM-SPS** [comment]
Hello @p-wysocki @jiwaszki @rkazants, is there any updates about this PR?

File: tools/mo/unit_tests/mo/back/ChangeRandomUniformOutputType_test.py
Comment (popovaan) [inline tools/mo/unit_tests/mo/back/ChangeRandomUniformOutputType_test.py:34]: Can we use @pytest.mark.parametrize instead of copying the test with different parameters?

File: tools/mo/unit_tests/mo/back/MatMulNormalizer_test.py
Comment (mvafin) [inline tools/mo/unit_tests/mo/back/MatMulNormalizer_test.py:30]: Why can't this be done using pytest functionality instead of for loop? Pytest is already used to run mo unit_tests in azure: https://github.com/openvinotoolkit/openvino/blob/c8bdefec307793acd51501e28ff0ac5294ceaf9b/.ci/azure/linux.yml#L462

File: tools/constraints.txt
Comment (rkazants) [inline tools/constraints.txt:18]: could you please remove `test-generator` from other `.txt` files: 
- tools\mo\requirements_dev.txt
- tests\samples_tests\smoke_tests\requirements.txt
- tests\constraints.txt

File: tools/mo/unit_tests/mo/back/ChangeRandomUniformOutputType_test.py
Comment (rkazants) [reply tools/mo/unit_tests/mo/back/ChangeRandomUniformOutputType_test.py:34]: @PRATHAM-SPS, can you please try to use pytests for tests that you are changing in your PR and use `parameterize` instead of tests-generator. Can you do this little experiment and share your experience?

Comment (PRATHAM-SPS) [reply tools/mo/unit_tests/mo/back/ChangeRandomUniformOutputType_test.py:34]: @rkazants @popovaan @mvafin I attempted to use pytest as recommended, but none of the test cases passed.

link:-  https://github.com/openvinotoolkit/openvino/compare/master...PRATHAM-SPS:openvino:pytest_testing?expand=1

Please advise if I'm doing it incorrectly.

Comment (rkazants) [reply tools/mo/unit_tests/mo/back/ChangeRandomUniformOutputType_test.py:34]: Hi @PRATHAM-SPS, I can give advice to take a look at our existing pytests here: https://github.com/openvinotoolkit/openvino/tree/master/tests/layer_tests

What kind or error are you getting?
I think the idea should be the same. Again. please update just one test and we will decide how to proceed.

@mvafin, @popovaan

Comment (popovaan) [reply tools/mo/unit_tests/mo/back/ChangeRandomUniformOutputType_test.py:34]: @PRATHAM-SPS Please try to apply my comments: https://github.com/openvinotoolkit/openvino/commit/3dadddf62e1d356a334786fbf15facf9c0f55069

With these corrections the test should work.

Comment (PRATHAM-SPS) [reply tools/mo/unit_tests/mo/back/ChangeRandomUniformOutputType_test.py:34]: @rkazants  E   ModuleNotFoundError: No module named 'openvino.tools.ovc' ,  how should i install it to resolve the issue? And yet another query? If pytest works properly, do I need to edit every file again? because it was suggested in the issue's description that it should be converted into a unittest's subTest, and it was confirmed that the file changes were acceptable.

Comment (popovaan) [reply tools/mo/unit_tests/mo/back/ChangeRandomUniformOutputType_test.py:34]: > ModuleNotFoundError: No module named 'openvino.tools.ovc'
Please merge the latest master. We removed OVC dependency from MO, so the issue should be fixed.

File: tools/constraints.txt
Comment (PRATHAM-SPS) [reply tools/constraints.txt:18]: @rkazants  Done, What I need to do next which task is ahead for me?

File: tools/mo/unit_tests/mo/back/ChangeRandomUniformOutputType_test.py
Comment (rkazants) [reply tools/mo/unit_tests/mo/back/ChangeRandomUniformOutputType_test.py:34]: @PRATHAM-SPS, did you experiment with using `pytest` just for one file instead of `generator`?

File: tools/constraints.txt
Comment (rkazants) [reply tools/constraints.txt:18]: @PRATHAM-SPS, the problem is that there appeared ovc unit tests that also use `generator`. Please temporarily recover `.txt` files with this module and work on replacing `generator` with `pytest`

**Comment by PRATHAM-SPS** [comment]
Hello @p-wysocki @rkazants @mvafin @jiwaszki @popovaan any update on this branch ??

**Comment by rkazants** [comment]
> Hello @p-wysocki @rkazants @mvafin @jiwaszki @popovaan any update on this branch ??

Hi @PRATHAM-SPS, could you please recover `generator` in requirements.txt file? Just let us have green CI.
After that, we would like to know your experience with replacing generator decorators with `pytest`. Just try to replace it in one file. May be it makes sense to experiment with this in separate PR, it is up to you.
What do you think?

Best regards,
Roman

File: tools/mo/unit_tests/mo/back/ChangeRandomUniformOutputType_test.py
Comment (PRATHAM-SPS) [reply tools/mo/unit_tests/mo/back/ChangeRandomUniformOutputType_test.py:34]: Hello @rkazants , pytest runs without an issue. One file was tested, and it passed every test case using Pytest

Here is the file:-https://github.com/openvinotoolkit/openvino/compare/master...PRATHAM-SPS:openvino:pytest_testing

and the results I posted right above your comment.

**Comment by PRATHAM-SPS** [comment]
> > Hello @p-wysocki @rkazants @mvafin @jiwaszki @popovaan any update on this branch ??
> 
> Hi @PRATHAM-SPS, could you please recover `generator` in requirements.txt file? Just let us have green CI. After that, we would like to know your experience with replacing generator decorators with `pytest`. Just try to replace it in one file. May be it makes sense to experiment with this in separate PR, it is up to you. What do you think?
> 
> Best regards, Roman

Good day, @rkazants.  Please check my new PR link for pytest, which I tried and provided to you above.

**Comment by rkazants** [comment]
> > > Hello @p-wysocki @rkazants @mvafin @jiwaszki @popovaan any update on this branch ??
> > 
> > 
> > Hi @PRATHAM-SPS, could you please recover `generator` in requirements.txt file? Just let us have green CI. After that, we would like to know your experience with replacing generator decorators with `pytest`. Just try to replace it in one file. May be it makes sense to experiment with this in separate PR, it is up to you. What do you think?
> > Best regards, Roman
> 
> Good day, @rkazants. Please check my new PR link for pytest, which I tried and provided to you above, its working alright. PR link:- [master...PRATHAM-SPS:openvino:pytest_testing](https://github.com/openvinotoolkit/openvino/compare/master...PRATHAM-SPS:openvino:pytest_testing)

@PRATHAM-SPS, can you please create pull request for this change (https://github.com/openvinotoolkit/openvino/compare/master...PRATHAM-SPS:openvino:pytest_testing)? It looks good to me and we will try to merge this new PR? Now I see `comparing changes` by link you provided:)

After this new PR merge, you will move step-by-step to avoid `test-generator` by the same replacing with pytest in other new PRs.

Best regards,
Roman

**Comment by PRATHAM-SPS** [comment]
Hello @rkazants @p-wysocki @popovaan @mvafin @jiwaszki  I have created a new PR with the above changes (Pytest). 
Link for New PR:- https://github.com/openvinotoolkit/openvino/pull/19298

**Comment by PRATHAM-SPS** [comment]
@rkazants @popovaan @p-wysocki @popovaan @mvafin I have a few doubts about this PR. I've been working on the assigned task for a month, but my profile doesn't reflect my progress. That commits i have made in this branch . Furthermore, if we close this PR without merging it, it will appear that I was inactive for the entire month. Is there a way to resolve this trouble?

(PROFILE ACTIVITY)
image

**Comment by PRATHAM-SPS** [comment]
@rkazants @p-wysocki @mvafin @jiwaszki @andrei-kochin  Do we need to close this PR?  : (

**Comment by mlukasze** [comment]
> @rkazants @p-wysocki @mvafin @jiwaszki @andrei-kochin Do we need to close this PR? : (

it may be, but your findings and changes will help us to make 2nd approach a lot faster and more accurate.
So it may that these exact changes need to be abandoned due change of approach, but we will use it as a guidance.
I understand that you are bounded to your work and it's highly appreciated - but sometimes correction on course need to be made.

**Comment by rkazants** [comment]
> @rkazants @p-wysocki @mvafin @jiwaszki @andrei-kochin Do we need to close this PR? : (

@PRATHAM-SPS, your second PR helped us understand that replacing with pytest is easy to do. And the solution gets much elegant than having `for` loop inside. Really sorry for that. we most likely have to close this PR.
Not sure if it works for you but you can override these changes with new approach in this PR or open new PR(s) instead of this one.

Best regards,
Roman

Best regards,
Roman

**Comment by PRATHAM-SPS** [comment]
> > Thanks @rkazants override will work for me. However, I'll do a separate PR for any files that I'm unsure either correct or incorrect.
> > 
> > But at the end, there will be 73 files to be reviewed all at once; is it okay with you to review them all at once?
> 
> I am fine to review 72 files if changes are quite similar. If you see some special case, you can decouple it into separate PR.

Thank you very much  for your consideration. I appreciate your understanding of the situation and your willingness to accommodate it :)

**Comment by PRATHAM-SPS** [comment]
Hello @rkazants @p-wysocki @popovaan @jiwaszki @mvafin @mlukasze @andrei-kochin , This PR is ready to merge , Now only 8 files are remaining I will cover that files in another PR as it contains special cases. Please review this PR and let me know if any changes are required.

File: tools/mo/unit_tests/mo/utils/freeze_placeholder_test.py
Comment (rkazants) [inline tools/mo/unit_tests/mo/utils/freeze_placeholder_test.py:160]: Please fix it here as well

Comment (PRATHAM-SPS) [reply tools/mo/unit_tests/mo/utils/freeze_placeholder_test.py]: Done this file comes under special test case thats why i have reverted it back to original one , I have already  created new PR for this one https://github.com/openvinotoolkit/openvino/pull/20313

**Review by rkazants — APPROVED** [review]

**Comment by rkazants** [comment]
build_jenkins

**Comment by PRATHAM-SPS** [comment]
Hello @rkazants  all CI are passed successfully.

**Pull request merged**
